### PR TITLE
Text: major cleanup in experimental component

### PR DESCRIPTION
**Pull request by dzearing**
The `Text` component represents a way to render themed text within an application. I went through and cleaned up a number of major issues blocking the usage of this component:

* Now respects the theme correctly.
* Defaults to the current font ramps.
* Renamed "type" to "variants" to work around potential future tslint problems.
* Ensure variants render correctly
* Adjusted theme creation logic a little to ensure typography is correctly processed
* Added a deep merge utility to reduce some of the potential for writing over the default typography settings.
* Refactored the text styling substantially
* Split style/view/types/registration files out into separated ones.

### CODE REVIEW

File: packages/experiments/src/components/Text/Text.styles.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  return {
+    root: [
+      typography.variants[variant!] as IStyle,
+      {
+        display: inline ? 'inline' : 'block'
+      },
+      family && {
+        // TODO: How are language specific font families configured?
+        fontFamily: typography.families[family]
```
Comment (natalieethell) [inline packages/experiments/src/components/Text/Text.styles.ts]: If this family doesn't exist in `typography.families`, what is used?

Comment (natalieethell) [reply packages/experiments/src/components/Text/Text.styles.ts]: Nevermind, with the defaults it shouldn't be an issue :)

File: packages/experiments/src/components/Text/Text.types.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+export interface ITextStyles {
+  /**
+   * Style for the root element.
+   */
+  root: IStyle;
+}
+
+// Inputs to the component
+export interface ITextProps extends IStyleableComponent<ITextProps, ITextStyles> {
+  as?: string | React.ReactType<ITextProps>;
```
Comment (natalieethell) [inline packages/experiments/src/components/Text/Text.types.tsx]: comment here?

File: packages/utilities/src/merge.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+/**
+ * Simple deep merge function. Takes all arguments are returns a new object which merges
+ * them all, in the order provided.
+ */
+export function merge<T = {}>(...args: Partial<T | null | undefined | false>[]): T {
```
Comment (natalieethell) [inline packages/utilities/src/merge.ts]: Cool!

Comment (cliffkoh) [inline packages/utilities/src/merge.ts]: Only `T` is a `Partial`

File: packages/utilities/src/merge.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+/**
+ * Simple deep merge function. Takes all arguments are returns a new object which merges
```
Comment (cliffkoh) [inline packages/utilities/src/merge.ts]: typo: and returns

File: packages/utilities/src/merge.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  return target as T;
+}
+
+function _merge<T>(target: T, source: T): T {
+  for (let name in source) {
+    if (source.hasOwnProperty(name)) {
+      const value = source[name];
+
+      if (typeof value === 'object') {
+        target[name] = merge(target[name], source[name]);
```
Comment (cliffkoh) [inline packages/utilities/src/merge.ts]: if there's a cyclic reference in the object, this will crash

File: packages/utilities/src/merge.ts (hunk 3 above)
Comment (cliffkoh) [inline packages/utilities/src/merge.ts]: The code supports arguments that are not all of the same type T but the typings do not have this flexibility. You should do something similar to the typings for mergeStyleSets/concatStyleSets

Comment (cliffkoh) [reply packages/utilities/src/merge.ts]: The BIG problem is that you cannot use `merge` without explicitly typing the shape of the fully merged object atm if some of the arguments are Partials of the final object, and it makes it really hard to use this utility function.

File: packages/utilities/src/merge.ts (hunk 5 above)
Comment (cliffkoh) [reply packages/utilities/src/merge.ts]: If a cycle is detected, this should not recurse further and instead just treat the object reference as any value, i.e. `target[name]=source[name];`

File: packages/experiments/src/components/Text/Text.styles.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import { IStyle, IPalette, ISemanticColors } from '../../Styling';
+import { IThemedProps } from '../../Foundation';
+import { ITextProps, ITextStyles } from './Text.types';
+
+export const TextStyles = (props: IThemedProps<ITextProps>): ITextStyles => {
```
Comment (cliffkoh) [inline packages/experiments/src/components/Text/Text.styles.ts]: is this the new "standard" for what is called `getStyles` everywhere in OUFR?

File: packages/experiments/src/components/Text/Text.tsx [hunk 7]
```diff
@@ -1,173 +1,13 @@
...
-      props.className
-    ]
-  } as ITextStyles;
-};
+import { createComponent } from '../../Foundation';
+import { ITextProps, ITextStyles } from './Text.types';
+import { TextView as view } from './Text.view';
+import { TextStyles as styles } from './Text.styles';
 
 export const Text: React.StatelessComponent<ITextProps> = createComponent<ITextProps, ITextStyles>({
```
Comment (cliffkoh) [inline packages/experiments/src/components/Text/Text.tsx]: This file should be a NEW file called Text.ts, so everything previously here is recognized as a rename to Text.base.ts and preserves history (presently, you will show up on the entire file's git blame and history as well)

File: packages/experiments/src/components/Text/Text.styles.ts (hunk 6 above)
Comment (dzearing) [reply packages/experiments/src/components/Text/Text.styles.ts]: agree! thanks

Comment (dzearing) [reply packages/experiments/src/components/Text/Text.styles.ts]: well.. what about react components?

```tsx
export const TextView = (props: IViewComponentProps<ITextProps, ITextStyles>) => {
```

Comment (dzearing) [reply packages/experiments/src/components/Text/Text.styles.ts]: Special case? Keep in mind also, you might decide to change TextStyles from an object to a function. Does that require a rename? Feels bad.

File: packages/experiments/src/components/Text/Text.styles.ts (hunk 1 above)
Comment (dzearing) [reply packages/experiments/src/components/Text/Text.styles.ts]: it would evaluate as undefined (and would render the default variant.)

**Comment by natalieethell** [comment]
Looks good to me! Only thing to note is that Katherine added some tests to theme.test.ts with the old typography.types name, which will just have to be changed to typography.variants.

**Pull request merged**